pkg/ssh/connection_golang.go

```diff
@@ -0,0 +1,355 @@
+	passwd, passwdSet := uri.User.Password()
+	if iden != "" { // iden might be blank if coming from image scp or if no validation is needed
+		value := iden
+		s, err := PublicKey(value, []byte(passwd))
+		if err != nil {
+			return nil, fmt.Errorf("failed to read identity %q: %w", value, err)
+		}
+		signers = append(signers, s)
+		logrus.Debugf("SSH Ident Key %q %s %s", value, ssh.FingerprintSHA256(s.PublicKey()), s.PublicKey().Type())
+	} else if sock, found := os.LookupEnv("SSH_AUTH_SOCK"); found { // validate ssh information, specifically the unix file socket used by the ssh agent.
+		logrus.Debugf("Found SSH_AUTH_SOCK %q, ssh-agent signer enabled", sock)
+
+		c, err := net.Dial("unix", sock)
+		if err != nil {
+			return nil, err
+		}
+		agentSigners, err := agent.NewClient(c).Signers()
+		if err != nil {
+			return nil, err
+		}
+
+		signers = append(signers, agentSigners...)
+
+		if logrus.IsLevelEnabled(logrus.DebugLevel) {
+			for _, s := range agentSigners {
+				logrus.Debugf("SSH Agent Key %s %s", ssh.FingerprintSHA256(s.PublicKey()), s.PublicKey().Type())
+			}
+		}
+	}
+	var authMethods []ssh.AuthMethod // now we validate and check for the authorization methods, most notaibly public key authorization
+	if len(signers) > 0 {
+		dedup := make(map[string]ssh.Signer)
+		for _, s := range signers {
+			fp := ssh.FingerprintSHA256(s.PublicKey())
+			if _, found := dedup[fp]; found {
+				logrus.Debugf("Dedup SSH Key %s %s", ssh.FingerprintSHA256(s.PublicKey()), s.PublicKey().Type())
+			}
+			dedup[fp] = s
+		}
+
+		var uniq []ssh.Signer
+		for _, s := range dedup {
+			uniq = append(uniq, s)
+		}
+		authMethods = append(authMethods, ssh.PublicKeysCallback(func() ([]ssh.Signer, error) {
+			return uniq, nil
+		}))
+	}
+	if passwdSet { // if password authentication is given and valid, add to the list
+		authMethods = append(authMethods, ssh.Password(passwd))
+	}
+	if len(authMethods) == 0 {
+		authMethods = append(authMethods, ssh.PasswordCallback(func() (string, error) {
+			pass, err := ReadPassword(fmt.Sprintf("%s's login password:", uri.User.Username()))
+			return string(pass), err
+		}))
+	}
+	tick, err := time.ParseDuration("40s")
+	if err != nil {
+		return nil, err
+	}
+	keyFilePath := filepath.Join(homedir.Get(), ".ssh", "known_hosts")
+	known, err := knownhosts.New(keyFilePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			f, err := os.OpenFile(keyFilePath, os.O_CREATE, 0o600)
+			if err != nil {
+				return nil, err
+			}
+			err = f.Close()
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	cfg := &ssh.ClientConfig{
+		User: uri.User.Username(),
+		Auth: authMethods,
+		HostKeyCallback: ssh.HostKeyCallback(func(host string, addr net.Addr, key ssh.PublicKey) error {
+			err := known(host, addr, key)
+			if err != nil {
+				keyErr := &knownhosts.KeyError{}
+				if errors.As(err, &keyErr) && len(keyErr.Want) > 0 {
+					// add the host. keyErr.Want len > 0 == there is a valid host key to be added
```

Comment:
I can’t see that it is OK to just silently use an unknown key. AFAICS if there _is_ a known_hosts entry, and the connection finds a key that does not match, this code path would trigger, ignore that mismatch, connect anyway, _and_ write that mismatching key. That’s fairly likely to be a compromised connection (or, sure, a reinstalled computer that lost its keys; there is no difference from the client’s point of view without human knowledge).

Reply:
If this can’t interactively ask the user at all, and there isn’t a `known_hosts` entry, I think the connection must fail before sending any credentials or anything else. (I could barely see an argument that it should proceed, but I feel fairly strongly against that.)

It _definitely_ must fail if there is a present and mismatching `known_hosts` entry.

---

If there is no known_hosts file, I don’t think we have to create one if we are not going to write anything to it (but we need to fail just as we would with an empty file.)

Reply:
that logic SGTM

Reply:
> If this can’t interactively ask the user at all, and there isn’t a known_hosts entry, I think the connection must fail before sending any credentials or anything else. 

Note that bootstrapping trust in this situation might require changes to tutorials (and `podman machine`, if it is going to use this code).

Reply:
i misinterpreted the docs: https://pkg.go.dev/golang.org/x/crypto/ssh/knownhosts#KeyError actually says when want == 0 that means you can add it, but when want > 0 that means it is a mismatch or unrecognized key